crates/data-structures/src/nstr.rs

```diff
@@ -0,0 +1,77 @@
+//! Provides UTF-8 strings of compile-time-known lengths.
+//!
+//! The strings can be constructed with `nstr!(string_literal)`
+//! producing a type `NStr<N>` where `const N: usize`.
+//! An example would be `nstr!("spacetime"): NStr<9>`.
+
+use core::{fmt, ops::Deref, str};
+
+/// A UTF-8 string of known length `N`.
```

Comment:
I'd want to be precise about what notion of length this is in the docs. I'd probably try to use language similar to what rust std uses when clarifying, eg in https://doc.rust-lang.org/std/string/struct.String.html#method.len